jme3-desktop/src/main/java/com/jme3/system/NativeLibraries.java

```diff
@@ -0,0 +1,170 @@
+ */
+package com.jme3.system;
+
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * Defines default native libraries used by jMonkeyEngine.
+ *
+ * @author Ali-RS
+ */
+public enum NativeLibraries {
+
+    // Note: LWJGL 3 handles its native library extracting & loading using
+    // its own SharedLibraryLoader.
+
+    // LWJGL 2
+    Lwjgl(new LibraryInfo("lwjgl", libPath ->
+            // lwjgl handle loading by itself.
+            System.setProperty("org.lwjgl.librarypath",
+                    Paths.get(libPath).getParent().toAbsolutePath().toString()))
+            .addNativeVariant(Platform.Windows32, "lwjgl.dll")
+            .addNativeVariant(Platform.Windows64, "lwjgl64.dll")
+            .addNativeVariant(Platform.Linux32, "liblwjgl.so")
+            .addNativeVariant(Platform.Linux64, "liblwjgl64.so")
+            .addNativeVariant(Platform.MacOSX32, "liblwjgl.dylib")
+            .addNativeVariant(Platform.MacOSX64, "liblwjgl.dylib")
+    ),
+
+    // OpenAL for LWJGL 2
+    // For OSX: Need to add lib prefix when extracting
+    Openal(new LibraryInfo("openal")
+            .addNativeVariant(Platform.Windows32, "OpenAL32.dll")
+            .addNativeVariant(Platform.Windows64, "OpenAL64.dll")
+            .addNativeVariant(Platform.Linux32,   "libopenal.so")
+            .addNativeVariant(Platform.Linux64,   "libopenal64.so")
+            .addNativeVariant(Platform.MacOSX32,  "openal.dylib", "libopenal.dylib")
+            .addNativeVariant(Platform.MacOSX64,  "openal.dylib", "libopenal.dylib")
+    ),
+
+    // BulletJme
+    BulletJme(new LibraryInfo("bulletjme")
+            .addNativeVariant(Platform.Windows32, "native/windows/x86/bulletjme.dll", "bulletjme-x86.dll")
+            .addNativeVariant(Platform.Windows64, "native/windows/x86_64/bulletjme.dll", "bulletjme-x86_64.dll")
+            .addNativeVariant(Platform.Windows_ARM64, "native/windows/arm64/bulletjme.dll", "bulletjme-arm64.dll")
+            .addNativeVariant(Platform.Linux32, "native/linux/x86/libbulletjme.so", "libbulletjme-x86.so")
+            .addNativeVariant(Platform.Linux64, "native/linux/x86_64/libbulletjme.so", "libbulletjme-x86_64.so")
+            .addNativeVariant(Platform.Linux_ARM32, "native/linux/arm32/libbulletjme.so", "libbulletjme-arm32.so")
+            .addNativeVariant(Platform.Linux_ARM64, "native/linux/arm64/libbulletjme.so", "libbulletjme-arm64.so")
+            .addNativeVariant(Platform.MacOSX32, "native/osx/x86/libbulletjme.dylib", "libbulletjme-x86.dylib")
+            .addNativeVariant(Platform.MacOSX64, "native/osx/x86_64/libbulletjme.dylib", "libbulletjme-x86_64.dylib")
+            .addNativeVariant(Platform.MacOSX_ARM64, "native/osx/arm64/libbulletjme.dylib", "libbulletjme-arm64.dylib")
+    ),
+
+    // JInput
+    // For OSX: Need to rename extension jnilib -> dylib when extracting
+    JInput(new LibraryInfo("jinput", libPath ->
```

Comment:
I don't understand what should the consumer loadFunction provide here ? other than the system load, but if the jinput handles the loading themselves, why would we even care about adding a loader function ?

Reply:
From what I understand both jinput and lwjgl2 do not support extracting the natives from the jar. They expect that the natives are extracted by application and the path to natives  is set via a system property. If the `net.java.games.input.librarypath` property is set then jinput will load them from there else it will use `System.loadLibrary()` to load from system-installed libraries.

I could not come up with a better solution so I added the load function style for custome load handling.

Reply:
LWJGL3 also uses a similar load function but for different purpose I guess

https://github.com/LWJGL/lwjgl3/blob/b44ba7af8030200bee34d7dfb77edf5617589328/modules/lwjgl/core/src/main/java/org/lwjgl/system/SharedLibraryLoader.java#L56-L60

Reply:
Hello there @Ali-RS,
Which class has the `System.load` to load the native libraries ?

Reply:
Sorry, I do not understand the question

Reply:
Nope, i mean the jme3 load function that loads the native libaries ?

Reply:
My original objection was on the loadFunction consumer name, the code uses some custom loading consumers which aren't actually real loading function, so i guess this is a kinda of injecting an action into a command-state pattern, since, the LibraryInfo has no idea of what the actual assigned function is supposed to do, i suggest renaming all the occurrences of the `loadFunction` to `onLoad` and documenting that this is called on the behalf of the `NativeLibraryLoader#loadNativeLibrary(..., ...)`.

Here is a list of good names, pick what you think is best: 
- onLoad
- nativeLoaderCallback
- nativeLoaderConsumer

Reply:
The `onLoad` may be also a little bit vague; because it is a listener for the caller of the stack `NativeLibraryLoader#loadNativeLibrary(..., ...)` and it is not a listener of the actual `System.load(String)`, but you can also consider the name `onExtractionCompleted` for example too, since you are extracting before loading and loading can be considered a listener, i haven't glanced the full code, so i cannot fully decide, but you already know most of the "under-the-hood" work, so i think you will select the best in this case, here is the list again: 
- onLoad
- nativeLoaderCallback
- nativeLoaderConsumer
- onExtractionCompleted

Reply:
I guess onExtractionCompleted will be the best, it is fully descriptive of the actual pattern, but idk, we don't extract every time we try to load a library too.......

Reply:
>and documenting that this is called on the behalf of the NativeLibraryLoader#loadNativeLibrary(..., ...).

Note NativeLibraryLoader#loadNativeLibrary(..., ...) internally calls the `loadFunction` of NativeLibrary it does not call System.load() directly.

Reply:
Yup, i know, it calls the consumer function...

Reply:
> ) or do something else.

And so that why i gave my objection to the name `loadFunction`.

Reply:
so I believe the name `loadFunction` suits more closely of what it does. (i.e it is not a listener)

Reply:
I see, no problem.